src/Network/Broadcast/OutboundQueue.hs

```diff
@@ -1032,32 +1044,33 @@ flush OutQ{..} = liftIO $ do
   queue again.
 -------------------------------------------------------------------------------}
 
--- | Update the set of peers
---
--- NOTE: Any messages to removed peers will be deleted from the queue.
+newtype Bucket nid = Bucket { bucketMVar :: MVar (Peers nid) }
+
+-- | Internal method: read all buckets of peers
+getAllPeers :: MonadIO m => OutboundQ msg nid buck -> m (Peers nid)
+getAllPeers OutQ{..} = liftIO $
+    mconcat <$> mapM (readMVar . bucketMVar) (Map.elems qBuckets)
+
+-- | Update a bucket of peers
 --
--- TODO check for overlap in the Peers terms. This should never happen but
--- could, due to misconfiguration.
-updateKnownPeers :: MonadIO m
-                 => OutboundQ msg nid -> (Peers nid -> Peers nid) -> m ()
-updateKnownPeers OutQ{..} f = do
-    removed <- applyMVar qPeers $ \peers ->
-      let peers'  = f peers
-          removed = peersToSet peers Set.\\ peersToSet peers'
-      in (peers', removed)
+-- NOTE: Any messages to peers that no longer exist in _any_ bucket will be
+-- removed from the queue.
+updatePeersBucket :: MonadIO m
+                  => OutboundQ msg nid buck
+                  -> buck
+                  -> (Peers nid -> Peers nid)
+                  -> m ()
+updatePeersBucket outQ@OutQ{..} buck f = do
+    before <- getAllPeers outQ
+    let Bucket bucket = qBuckets Map.! buck
+    applyMVar_ bucket f
+    after <- getAllPeers outQ
+    let removed = peersToSet before Set.\\ peersToSet after
```

Comment:
> it's not entirely clear what the final outcome will be anyway

Depending on which thread enters `updatePeersBucket` first: 

- The node is added (re-added)  to the buckets, and then subsequently removed by the other thread.
- The node is *completely* removed (relevant updates to `qInFlight`, `qFailures`, and the `MQ` finish), and the node is subsequently re-added